### ISSUE TITLE
`DEPTHWISE_CONV_2D` with `depth_multiplier > 1` is ~860x slower than the equivalent op via ONNX→QNN on the same device

### ISSUE THREAD

**Issue by airyym** [google-ai-edge/litert-torch#1211]
## Summary

A `DEPTHWISE_CONV_2D` with `depth_multiplier > 1` runs orders of magnitude slower on the
QNN/HTP delegate than the same computation reached through other paths.

A single `Conv2d(64, 128, kernel=7, stride=2, groups=64)` layer takes **459 ms** on the NPU
with litert-torch 0.9.4 (230 ms on 0.8.0). For comparison, on the same device:

- the same layer with `depth_multiplier == 1`: **0.68 ms**
- the same layer as a grouped conv (`groups=32`, not depthwise): **0.69 ms**
- the identical computation via ONNX → ORT-QNN EP, same HTP backend: **0.269 ms**
- the same layer on CPU: **2.7 ms** (i.e. the NPU is ~170x slower than CPU here)

Since the ONNX→QNN path executes this efficiently on the same hardware, the cost does not
look like an HTP capability limit. Below is what we observed in the emitted QNN graph; we
have not been able to confirm the mechanism.

## Environment

| | |
|---|---|
| Device | Snapdragon 8 Elite / SM8750, Hexagon **v79** |
| LiteRT | **2.1.6** (`com.google.ai.edge.litert:litert`) |
| QNN runtime | **QAIRT 2.47.0** (`qualcomm_runtime_v79`) |
| Converter | `litert-torch` **0.9.4** (latest; also reproduced on 0.8.0) |
| Accelerator | `Accelerator.NPU` (QNN HTP delegate) |
| Options | `htpPerformanceMode = BURST`, `optimizationLevel = HTP_OPTIMIZE_FOR_INFERENCE_O3`, `useConvHmx = true`, `vtcmSize = 8` |
| Model precision | fp32 `.tflite` (plugin executes it as fp16 on HTP) |

## Reproduction

```python
import torch, litert_torch

class DW(torch.nn.Module):
    def __init__(self, cin, cout):
        super().__init__()
        # groups == cin  → depthwise.  cout // cin == depth_multiplier
        self.c = torch.nn.Conv2d(cin, cout, 7, stride=2, padding=3, groups=cin, bias=False)
    def forward(self, x):
        return self.c(x)

x = torch.randn(1, 64, 80, 80)
litert_torch.convert(DW(64,  64).eval(), (x,)).export("dw_mult1.tflite")  # fast
litert_torch.convert(DW(64, 128).eval(), (x,)).export("dw_mult2.tflite")  # ~320x slower
```

Run both with `Accelerator.NPU` and the options above; warmup 5, 30 iterations, median.

## Observation: weight tensor shape in the emitted QNN graph

QNN graph dumped via `QualcommOptions.irJsonDir`:

| | `depth_multiplier = 1` (fast) | `depth_multiplier = 2` (slow) |
|---|---|---|
| input | `[1, 80, 80, 64]` | `[1, 80, 80, 64]` |
| **weight** | `[7, 7, 1, 64]` | **`[7, 7, 1, 128]`** |
| **expected** | — | **`[7, 7, 64, 2]`** |
| output | `[1, 40, 40, 64]` | `[1, 40, 40, 128]` |

In the slow case the weight is `[7, 7, 1, 128]`, which looks like the TFLite weight
`[1, 7, 7, 128]` with axes permuted and no channel/multiplier split applied — the channel
axis stays 1 while all 128 output channels sit in the last axis, even though the input
carries 64 channels.

We are not certain what QNN expects here (we could not find the `DepthWiseConv2d` weight
layout documented), so `[7, 7, 64, 2]` above is our guess at the intended form rather than
a verified requirement. It is possible this shape is legal and the slowdown has another
cause.

Everything else in the emitted graph looks unremarkable: a single `DepthWiseConv2d` node,
one partition with no CPU fallback, and `stride` / `dilation` / `pad_amount` all matching
the source op. The weight shape is the only difference we could find between the fast and
slow cases.

This matches the single line in
[`Qualcomm_QNN_Compiler.md`](https://github.com/google-ai-edge/LiteRT/blob/main/litert/vendors/qualcomm/compiler/Qualcomm_QNN_Compiler.md):

> `kLiteRtOpCodeTflDepthwiseConv2d` | Legalized to `QNN_OP_DEPTH_WISE_CONV_2D`.
> Supports fused activation. **Weights are reshaped.**

If that reshape is meant to fold `depth_multiplier` into the channel axis, it does not
appear to happen here. TFLite-side weights for the two cases above are `[1, 7, 7, 64]` and
`[1, 7, 7, 128]`, so `in_channels` (from the input tensor) and
`multiplier = out_channels / in_channels` are both derivable at legalization time.

### Reproduced on both converter versions (on-device, median of 30)

| Probe | litert-torch 0.8.0 | **litert-torch 0.9.4** | CPU (either) |
|---|---|---|---|
| multiplier 1 (64→64) | 0.72 ms | **0.68 ms** | 2.3 ms |
| **multiplier 2 (64→128)** | 230.77 ms | **459.56 ms** | 2.7 ms |
| **multiplier 4 (64→256)** | 460.36 ms | **909.24 ms** | 4.1 ms |

0.9.4 folds the convolution padding into the input tensor (`[1, 86, 86, 64]`, no separate
`Pad` node) whereas 0.8.0 emits an explicit `Pad`; we did not isolate whether that graph
difference contributes to the 2x. The tables below are from 0.8.0.

### LiteRT QNN delegate (on-device, median of 30)

| Probe | Shape | `depth_multiplier` | **NPU** | CPU |
|---|---|---|---|---|
| `m1_64to64` | 64→64, groups=64, k7 s2 | 1 | **0.71 ms** | 2.18 ms |
| `m2_64to128` | 64→128, groups=64, k7 s2 | **2** | **230.24 ms** | 2.70 ms |
| `m4_64to256` | 64→256, groups=64, k7 s2 | **4** | **460.36 ms** | 3.72 ms |
| `m2_s1_64to128` | 64→128, groups=64, k7 **s1** | 2 | **913.90 ms** | 7.72 ms |
| `m2_16to32` | 16→32, groups=16, k7 s2 | 2 | 116.11 ms | 1.40 ms |
| `m2_8to16` | 8→16, groups=8, k7 s2 | 2 | 58.90 ms | 1.08 ms |
| `g32_64to128` | 64→128, **groups=32** (grouped, not depthwise) | — | **0.69 ms** | 2.10 ms |

### Same computation via ONNX → ORT-QNN EP (same device, same HTP backend)

Exported with `torch.onnx.export(..., opset_version=17)`; run through onnxruntime-qnn with
`backend_path=libQnnHtp.so`, `enable_htp_fp16_precision=1`, `htp_performance_mode=burst`.
Median of 30, `session.Run` only.

| Probe | ONNX → ORT-QNN | LiteRT → QNN | ratio |
|---|---|---|---|
| multiplier 1 (64→64) | 0.237 ms | 0.71 ms | 3x |
| **multiplier 2 (64→128)** | **0.269 ms** | **230.24 ms** | **856x** |
| **multiplier 4 (64→256)** | 0.347 ms | 460.36 ms | 1327x |
| multiplier 2, stride 1 | 0.512 ms | 913.90 ms | 1785x |
| grouped (groups=32) | 0.275 ms | 0.69 ms | 2.5x |

**Comment by aravindmurali22** [comment]
Hi @YunanAZ , following up on the thread. Is there any update on the issue? Thank you.

**Comment by qti-hungjuiw** [comment]
Hi @airyym, @aravindmurali22,
1. The root cause is that **Conv2d** can be optimized more effectively on QNN HTP than **DepthwiseConv2d**. During `torch.onnx.export`, **DepthwiseConv2d** is converted to **Conv2d**, which is why you are seeing significantly better performance.
2. We have a draft PR that validates this behavior, although we currently do not plan to merge it. You can either cherry-pick the changes from the PR or directly replace **DepthwiseConv2d** with **Conv2d** (using **groups=in_channels**) in the source model to achieve the same performance improvement.
    - https://github.com/google-ai-edge/LiteRT/pull/9931
